src/core/math/quat.js

```diff
@@ -461,103 +476,59 @@ class Quat {
+        let m01 = d[1];
+        let m02 = d[2];
+        let m10 = d[4];
+        let m11 = d[5];
+        let m12 = d[6];
+        let m20 = d[8];
+        let m21 = d[9];
+        let m22 = d[10];
+
+        let l;
+
+        // remove scaling from axis vectors
+        l = m00 * m00 + m01 * m01 + m02 * m02;
+        if (l === 0) return this.set(0, 0, 0, 1);
+        l = 1 / Math.sqrt(l);
+        m00 *= l;
+        m01 *= l;
+        m02 *= l;
+
+        l = m10 * m10 + m11 * m11 + m12 * m12;
+        if (l === 0) return this.set(0, 0, 0, 1);
+        l = 1 / Math.sqrt(l);
+        m10 *= l;
+        m11 *= l;
+        m12 *= l;
+
+        l = m20 * m20 + m21 * m21 + m22 * m22;
+        if (l === 0) return this.set(0, 0, 0, 1);
+        l = 1 / Math.sqrt(l);
+        m20 *= l;
+        m21 *= l;
+        m22 *= l;
+
+        // https://d3cw3dd2w32x2b.cloudfront.net/wp-content/uploads/2015/01/matrix-to-quat.pdf
+
+        if (m22 < 0) {
             if (m00 > m11) {
-                if (m00 > m22) {
-                    // XDiagDomMatrix
-                    rs = (m00 - (m11 + m22)) + 1;
-                    rs = Math.sqrt(rs);
-
-                    this.x = rs * 0.5;
-                    rs = 0.5 / rs;
-                    this.w = (m12 - m21) * rs;
-                    this.y = (m01 + m10) * rs;
-                    this.z = (m02 + m20) * rs;
-                } else {
-                    // ZDiagDomMatrix
-                    rs = (m22 - (m00 + m11)) + 1;
-                    rs = Math.sqrt(rs);
-
-                    this.z = rs * 0.5;
-                    rs = 0.5 / rs;
-                    this.w = (m01 - m10) * rs;
-                    this.x = (m20 + m02) * rs;
-                    this.y = (m21 + m12) * rs;
-                }
-            } else if (m11 > m22) {
-                // YDiagDomMatrix
-                rs = (m11 - (m22 + m00)) + 1;
-                rs = Math.sqrt(rs);
-
-                this.y = rs * 0.5;
-                rs = 0.5 / rs;
-                this.w = (m20 - m02) * rs;
-                this.z = (m12 + m21) * rs;
-                this.x = (m10 + m01) * rs;
+                this.set(1 + m00 - m11 - m22, m01 + m10, m20 + m02, m12 - m21);
             } else {
-                // ZDiagDomMatrix
-                rs = (m22 - (m00 + m11)) + 1;
-                rs = Math.sqrt(rs);
-
-                this.z = rs * 0.5;
-                rs = 0.5 / rs;
-                this.w = (m01 - m10) * rs;
-                this.x = (m20 + m02) * rs;
-                this.y = (m21 + m12) * rs;
+                this.set(m01 + m10, 1 - m00 + m11 - m22, m12 + m21, m20 - m02);
+            }
+        } else {
+            if (m00 < -m11) {
+                this.set(m20 + m02, m12 + m21, 1 - m00 - m11 + m22, m01 - m10);
+            } else {
+                this.set(m12 - m21, m20 - m02, m01 - m10, 1 + m00 + m11 + m22);
             }
         }
 
-        return this;
+        return this.mulScalar(1.0 / this.length());
```

Comment:
Maybe put a comment about why your need to renormalize?

Reply:
done